capellambse/auditing.py

```diff
@@ -0,0 +1,86 @@
+# Copyright 2021 DB Netz AG
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+import collections.abc as cabc
+import sys
+import typing as t
+
+import capellambse
+
+from . import helpers
+
+
+class AttributeAuditor:
+    """Audits access to attributes of ModelElements.
+
+    .. warning::
+         This will permanently add an audit hook to the global hook
+        table. The auditor will keep the model alive, which may consume
+        excessive memory. To avoid this, call the auditor object's
+        ``detach()`` method once you are done with it. This is
+        automatically done if you use it as a context manager.
+
+    Examples
+    --------
+    >>> auditor = AttributeAuditor(model, {"name", "description"})
+    >>> print(model.la.all_components[0].name)
+    Hogwarts
+    >>> auditor.recorded_ids
+    {'0d2edb8f-fa34-4e73-89ec-fb9a63001440'}
+    >>> # Cleanup
+    >>> auditor.model = None
+
+    >>> with AttributeAuditor(model, {"name", "description"}) as auditor:
+    ...     print(model.la.all_components[0].name)
+    ...
+    Hogwarts
+    >>> auditor.recorded_ids
+    {'0d2edb8f-fa34-4e73-89ec-fb9a63001440'}
```

Comment:
What do you think about changing the interface of the context manager so that it directly gives the set of `recorded_ids` to the `as` keyword? The only reason that I can see why one would need the auditor object is to `detach()` it afterwards, which is done by the context manager automatically anyways. Directly handing over the set of IDs makes the usage a little more straight forward.